servicetalk-opentracing-zipkin-publisher/src/main/java/io/servicetalk/opentracing/zipkin/publisher/reporter/UdpReporter.java

```diff
@@ -0,0 +1,252 @@
+
+        /**
+         * Builds a new {@link UdpReporter} instance with this builder's options.
+         * <p>
+         * This method may block while the underlying UDP channel is being bound.
+         *
+         * @return a new {@link UdpReporter}
+         */
+        public UdpReporter build() {
+            return new UdpReporter(this);
+        }
+    }
+
+    private Bootstrap buildBootstrap(EventLoopGroup group, Codec codec, SocketAddress collectorAddress,
+                                     @Nullable String loggerName) {
+        if (!(collectorAddress instanceof InetSocketAddress)) {
+            throw new IllegalArgumentException("collectorAddress " + collectorAddress +
+                    " is invalid for UDP");
+        }
+        return new Bootstrap()
+                .group(group)
+                .channel(datagramChannel(group))
+                .option(RCVBUF_ALLOCATOR, DEFAULT_RECV_BUF_ALLOCATOR)
+                .handler(new ChannelInitializer<Channel>() {
+                    @Override
+                    protected void initChannel(final Channel ch) {
+                        if (loggerName != null) {
+                            ch.pipeline().addLast(new LoggingHandler(loggerName, LogLevel.TRACE));
+                        }
+                        ch.pipeline().addLast(new ChannelOutboundHandlerAdapter() {
+                            @Override
+                            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                                if (msg instanceof Span) {
+                                    byte[] bytes = codec.encoder.encode((Span) msg);
+                                    ByteBuf buf = ctx.alloc().buffer(bytes.length).writeBytes(bytes);
+                                    ctx.write(new DatagramPacket(buf, (InetSocketAddress) collectorAddress), promise);
+                                } else {
+                                    ctx.write(msg, promise);
+                                }
+                            }
+                        });
+                    }
+                });
+    }
+
+    /**
+     * Non-blocking report method.
+     *
+     * @param span the span to report
+     */
+    @Override
+    public void report(final Span span) {
+        if (!channel.isActive()) {
+            throw new RuntimeException(StacklessClosedChannelException.newInstance(this.getClass(), "report"));
+        }
+        channel.writeAndFlush(span);
+    }
+
+    /**
+     * Blocking close method delegates to {@link #closeAsync()}).
+     */
+    @Override
+    public void close() {
+        awaitTermination(closeable.closeAsync().toFuture());
+    }
+
+    /**
+     * Marks the {@link UdpReporter} as closed to prevent accepting any more {@link Span}s.
+     *
+     * @return a {@link Completable} that is completed when close is done.
+     */
+    @Override
+    public Completable closeAsync() {
+        return closeable.closeAsync();
+    }
```

Comment:
Also override `closeAsyncGracefully()`

```suggestion
    }

    @Override
    public Completable closeAsyncGracefully() {
        return closeable.closeAsyncGracefully();
    }    
```